test/e2e/storage/external/external.go

```diff
@@ -258,39 +262,44 @@ func (d *driverDefinition) SkipUnsupportedTest(pattern testpatterns.TestPattern)
 }
 
 func (d *driverDefinition) GetDynamicProvisionStorageClass(config *testsuites.PerTestConfig, fsType string) *storagev1.StorageClass {
+	var (
+		sc  *storagev1.StorageClass
+		err error
+	)
+
 	f := config.Framework
 
-	if d.StorageClass.FromName {
-		provisioner := d.DriverInfo.Name
-		parameters := map[string]string{}
-		ns := f.Namespace.Name
-		suffix := provisioner + "-sc"
-		if fsType != "" {
-			parameters["csi.storage.k8s.io/fstype"] = fsType
-		}
+	switch {
+	case d.StorageClass.FromName:
+		sc = &storagev1.StorageClass{Provisioner: d.DriverInfo.Name}
+	case d.StorageClass.FromExistingClassName != "":
+		sc, err = f.ClientSet.StorageV1().StorageClasses().Get(d.StorageClass.FromExistingClassName, metav1.GetOptions{})
+		framework.ExpectNoError(err, "getting storage class %s", d.StorageClass.FromExistingClassName)
+	case d.StorageClass.FromFile != "":
+		var ok bool
 
-		return testsuites.GetStorageClass(provisioner, parameters, nil, ns, suffix)
-	}
+		items, err := utils.LoadFromManifests(d.StorageClass.FromFile)
+		framework.ExpectNoError(err, "load storage class from %s", d.StorageClass.FromFile)
+		framework.ExpectEqual(len(items), 1, "exactly one item from %s", d.StorageClass.FromFile)
+
+		err = utils.PatchItems(f, items...)
+		framework.ExpectNoError(err, "patch items")
 
-	items, err := utils.LoadFromManifests(d.StorageClass.FromFile)
-	framework.ExpectNoError(err, "load storage class from %s", d.StorageClass.FromFile)
-	framework.ExpectEqual(len(items), 1, "exactly one item from %s", d.StorageClass.FromFile)
+		sc, ok = items[0].(*storagev1.StorageClass)
+		framework.ExpectEqual(ok, true, "storage class from %s", d.StorageClass.FromFile)
+	}
 
-	err = utils.PatchItems(f, items...)
-	framework.ExpectNoError(err, "patch items")
+	framework.ExpectNotEqual(sc, nil, "storage class is unexpectantly nil")
 
-	sc, ok := items[0].(*storagev1.StorageClass)
-	framework.ExpectEqual(ok, true, "storage class from %s", d.StorageClass.FromFile)
-	// Ensure that we can load more than once as required for
-	// GetDynamicProvisionStorageClass by adding a random suffix.
-	sc.Name = names.SimpleNameGenerator.GenerateName(sc.Name + "-")
 	if fsType != "" {
 		if sc.Parameters == nil {
 			sc.Parameters = map[string]string{}
 		}
+		// This limits the external storage test suite to only CSI drivers, which may need to be
+		// reconsidered if we eventually need to move in-tree storage tests out.
 		sc.Parameters["csi.storage.k8s.io/fstype"] = fsType
 	}
-	return sc
+	return testsuites.GetStorageClass(sc.Provisioner, sc.Parameters, sc.VolumeBindingMode, f.Namespace.Name, "e2e-sc")
```

Comment:
When `FromName` is used, the provisioner is no longer part of the SC name (it was before).

This shouldn't break anything, though...

Reply:
Yeah I think it's ok. It's kind of a moot point now since we added the csi fstype param that makes this test suite csi-only, but if you did want to run this against an in-tree volume type, then the "~" in the in-tree provisioner name would cause storageclass creation to fail due to invalid character.